toolbox/component_tool.go

```diff
@@ -0,0 +1,213 @@
+package toolbox
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/cloudtrust/keycloak-client/v2"
+)
+
+var (
+	ErrConfigKeyNotFound   = errors.New("component config key not found")
+	ErrInvalidConfigFormat = errors.New("invalid config encoding")
+)
+
+type ComponentTool interface {
+	FindComponent(components []keycloak.ComponentRepresentation) *keycloak.ComponentRepresentation
+	InitializeComponent(realmName string, idpID string, initial any) (keycloak.ComponentRepresentation, error)
+	GetComponentEntry(comp *keycloak.ComponentRepresentation, key string, out any) error
+	UpdateComponentEntry(comp *keycloak.ComponentRepresentation, key string, value any) error
+	DeleteComponentEntry(comp *keycloak.ComponentRepresentation, key string) (bool, error)
+
+	GetProviderType() string
+}
+
+type ComponentConfig struct {
+	ProviderType string `mapstructure:"provider-type"`
+	ProviderID   string `mapstructure:"provider-id"`
+	ConfigName   string `mapstructure:"config-name"`
+}
+
+type GenericComponentTool struct {
+	ProviderType string
+	ProviderID   string
+	ConfigName   string
+}
+
+func NewComponentTool(config ComponentConfig) ComponentTool {
+	return &GenericComponentTool{
+		ProviderType: config.ProviderType,
+		ProviderID:   config.ProviderID,
+		ConfigName:   config.ConfigName,
+	}
+}
+
+type ComponentConfigKeyValue struct {
+	Key   string `json:"key"`
+	Value string `json:"value"`
+}
+
+func (ct *GenericComponentTool) GetProviderType() string {
+	return ct.ProviderType
+}
+
+func (ct *GenericComponentTool) FindComponent(components []keycloak.ComponentRepresentation) *keycloak.ComponentRepresentation {
+	for i := range components {
+		c := &components[i]
+		if c.ProviderID != nil && *c.ProviderID == ct.ProviderID {
+			return c
+		}
+	}
+
+	return nil
+}
+
+func (ct *GenericComponentTool) InitializeComponent(realmName string, idpID string, initial any) (keycloak.ComponentRepresentation, error) {
+	valueJSON, err := json.Marshal(initial)
+	if err != nil {
+		return keycloak.ComponentRepresentation{}, err
+	}
+
+	arr := []ComponentConfigKeyValue{{
+		Key:   idpID,
+		Value: string(valueJSON),
+	}}
+
+	arrJSON, err := json.Marshal(arr)
+	if err != nil {
+		return keycloak.ComponentRepresentation{}, err
+	}
+
+	config := map[string][]string{
+		ct.ConfigName: {string(arrJSON)},
+	}
+
+	return keycloak.ComponentRepresentation{
+		ProviderID:   &ct.ProviderID,
+		ProviderType: &ct.ProviderType,
+		ParentID:     &realmName,
+		Config:       &config,
+	}, nil
+}
+
+func (ct *GenericComponentTool) loadConfigArray(comp *keycloak.ComponentRepresentation) ([]ComponentConfigKeyValue, error) {
+	if comp == nil || comp.Config == nil {
+		return nil, fmt.Errorf("no config found")
+	}
+	entries := (*comp.Config)[ct.ConfigName]
```

Comment:
That's why map and slices should never be declared with pointers in "Representations"